### ISSUE TITLE
Add an option to sort added tracks by configurable script, after them being added to playlist

### ISSUE THREAD

**Issue by orbea** [Alexey-Yakovenko/deadbeef#1560]
OS: `Slackware64-current`
`deadbeef-8214e4f_2016.04.22_master-x86_64-1_git`

When adding folders (albums) to a playlist the tracks are added in the order of the file names rather than being sorted by the track number included in the metadata as in music players such as cmus or fb2k.

In the below example the albums `Monotony Fields` and `Shades Of` are added in the correct order while `Illusion's Play` and `angels of distress` are added in the incorrect track order.  This is especially annoying as the sort option is not capable of sorting only the highlighted tracks as in fb2k and requires the albums to be sorted by hand each time they are added to a playlist.
```
Shape Of Despair$ ls *
Illusion's Play:
Curse Life.flac          Fragile Emptiness.flac  Sleep Mirrored.flac
Entwined in Misery.flac  Illusion's Play.flac    Still-motion.flac

Monotony Fields:
Shape Of Despair - Monotony Fields - 01 Reaching The Innermost.flac
Shape Of Despair - Monotony Fields - 02 Monotony Fields.flac
Shape Of Despair - Monotony Fields - 03 Descending Inner Night.flac
Shape Of Despair - Monotony Fields - 04 The Distant Dream Of Life.flac
Shape Of Despair - Monotony Fields - 05 Withdrawn.flac
Shape Of Despair - Monotony Fields - 06 In Longing.flac
Shape Of Despair - Monotony Fields - 07 The Blank Journey.flac

Shades Of:
01 - ...In The Mist.flac  03 - Down Into The Stream.flac  05 - Sylvan-Night.flac
02 - Woundheir.flac       04 - Shadowed Dreams.flac

angels of distress:
Angels Of Distress.flac  Quiet These Paintings Are.flac
Fallen.flac              To Live For My Death.flac
Night's Dew.flac
```

**Comment by Alexey-Yakovenko** [comment]
This is by design, and not going to be changed. Changing this to be sorted by metadata would break far more use cases, than fix.

**Comment by orbea** [comment]
Out of curiosity do you have an example of a use case this would break? Even if this behavior is not added there are less drastic ways of "solving" this. Like if it were possible to sort only highlighted tracks in a playlist than the entire playlist.

**Comment by Alexey-Yakovenko** [comment]
> Out of curiosity do you have an example of a use case this would break?

Yeah.. how would you sort this? I couldn't make the sorting work. There are many other corner cases like this, which work fine when sorting by %path%, but would require special code if you want to sort by metadata.

![image](https://cloud.githubusercontent.com/assets/1895698/15019031/830bbbae-121c-11e6-88cb-5cdffb0adeda.png)

**Comment by Alexey-Yakovenko** [comment]
> Like if it were possible to sort only highlighted tracks in a playlist than the entire playlist.

This issue is asking to sort the files by metadata, when they're being added to playlist. If you want to request another feature -- e.g. sorting selected tracks -- please post a new feature request.

**Comment by Alexey-Yakovenko** [comment]
> Would it still break use cases if this was an optional feature so that users could choose how they want tracks sorted by default?

I closed this because of the way this issue is written - as a bug. It implies changing the way files are sorted from their natural (fs-based) sorting to post-process sorting, based on metadata. This is not going to happen.

Rewording this issue into an feature request to add an option to sort added files after adding them to playlist -- would work better.

**Comment by Alexey-Yakovenko** [comment]
@orbea 

I'm open to suggestions which title formatting to use for default post-process sorting.

I'd start with something like this:

```
$directory_path(%path%)
[%disc%]
%track number%
%album%
%artist%
```

**Comment by Alexey-Yakovenko** [comment]
another script idea

```
$directory_path(%path%)-
$if2($pad(%disc%,2,0),00)-
$if2(%track number%,00)-
[%album%]-
[%artist%]
```

This might even work.

**Comment by orbea** [comment]
I was not sure on how you thought would be the best way to address this so I left my wording possibly vague.

The problem with my use case is that my music library is extensive and the CDs I ripped years ago don't have the track number in the file name. So this left my music library with inconsistent naming conventions which would take a lot of work to untangle. This has not ever been a problem with other players (cmus, fb2k) because they relied exclusively on the metadata defaulting to the file name only if the metadata was missing.

With deadbeef it adds tracks with only the file name, so that it fails with a huge portion of my music library leaving me to sort albums manually by hand. This is a minor and constant inconvenience except for the few albums with 20+ tracks where it becomes more work to sort than reasonable.

My ideal solution would be to add two features.
* To be able to configure the default behavior for sorting any new tracks added, preferably to address all use cases.
* To be able to sort only select tracks in a playlist by hilighting them and then using the sort function which could also be added to the right click menu. Fb2k implements this feature and could be used as an example for how it should work. (This would also be very helpful for other use cases too, like to organize large playlists.)

I did not think of the script idea when I made this issue report, I'm not really sure how effectively it would solve my personal use case yet. Though I am certainly open to trying it. :)

**Comment by Alexey-Yakovenko** [comment]
> I did not think of the script idea

That's the only way how sorting works in deadbeef - it's always a title formatting script.

**Comment by niedzielski** [comment]
I would love a way to sort files that were just added to a playlist without sorting the entire playlist.

My usual workflow is:

1. I have a long playlist and I'm far into it.
2. I append an album.
3. The newly added album is not ordered by track number.

At that point, my options seem to be:

A. Sort the entire playlist.
B. Manually reorder the newly added album's tracks.

Option A technically works. I have a column configured to sort by `%path%`:

<img width="597" height="524" alt="Image" src="https://github.com/user-attachments/assets/a24a32a1-645c-4788-ab5c-f31f3c3e7c17" />

However, I almost never use option A because it reorders the whole playlist and makes me lose my place. That leaves only option B.

That's where an option C would be ideal: sort only the newly added files. Maybe if there was a way to set a default sort order, such as `%path%`, this could be used for operations like append.

For example, here is the original ordering of an appended album:

```text
file:///home/user/audio/Sakanaction/2019%20834.194/2-1%20Good-Bye.m4a
file:///home/user/audio/Sakanaction/2019%20834.194/1-1%20Wasurerarenaino.m4a
file:///home/user/audio/Sakanaction/2019%20834.194/2-2%20Hasu%20no%20Hana%20(Single%20Version).m4a
file:///home/user/audio/Sakanaction/2019%20834.194/1-2%20Match%20to%20Peanuts.m4a
file:///home/user/audio/Sakanaction/2019%20834.194/2-3%20Eureka.m4a
file:///home/user/audio/Sakanaction/2019%20834.194/1-3%20Kagerou.m4a
file:///home/user/audio/Sakanaction/2019%20834.194/2-4%20Nylon%20no%20Ito.m4a
file:///home/user/audio/Sakanaction/2019%20834.194/1-4%20Tabun,%20Kaze..m4a
file:///home/user/audio/Sakanaction/2019%20834.194/2-5%20Chabashira.m4a
file:///home/user/audio/Sakanaction/2019%20834.194/1-5%20Shin%20Takarajima.m4a
file:///home/user/audio/Sakanaction/2019%20834.194/1-6%20Moth.m4a
file:///home/user/audio/Sakanaction/2019%20834.194/2-6%20Wonderland.m4a
file:///home/user/audio/Sakanaction/2019%20834.194/1-7%20_Kikitakatta%20Dance%20Music,%20Liquidroom%20ni_.m4a
file:///home/user/audio/Sakanaction/2019%20834.194/2-7%20Sayonara%20wa%20Emotion.m4a
file:///home/user/audio/Sakanaction/2019%20834.194/2-8%20834.194.m4a
file:///home/user/audio/Sakanaction/2019%20834.194/1-8%20Eureka%20(Shotaro%20Aoyama%20Remix).m4a
file:///home/user/audio/Sakanaction/2019%20834.194/2-9%20September%20(Sapporo%20Version).m4a
file:///home/user/audio/Sakanaction/2019%20834.194/1-9%20September%20(Tokyo%20Version).m4a
```

Here's the desired order which I generated using option A:

```text
file:///home/user/audio/Sakanaction/2019%20834.194/1-1%20Wasurerarenaino.m4a
file:///home/user/audio/Sakanaction/2019%20834.194/1-2%20Match%20to%20Peanuts.m4a
file:///home/user/audio/Sakanaction/2019%20834.194/1-3%20Kagerou.m4a
file:///home/user/audio/Sakanaction/2019%20834.194/1-4%20Tabun,%20Kaze..m4a
file:///home/user/audio/Sakanaction/2019%20834.194/1-5%20Shin%20Takarajima.m4a
file:///home/user/audio/Sakanaction/2019%20834.194/1-6%20Moth.m4a
file:///home/user/audio/Sakanaction/2019%20834.194/1-7%20_Kikitakatta%20Dance%20Music,%20Liquidroom%20ni_.m4a
file:///home/user/audio/Sakanaction/2019%20834.194/1-8%20Eureka%20(Shotaro%20Aoyama%20Remix).m4a
file:///home/user/audio/Sakanaction/2019%20834.194/1-9%20September%20(Tokyo%20Version).m4a
file:///home/user/audio/Sakanaction/2019%20834.194/2-1%20Good-Bye.m4a
file:///home/user/audio/Sakanaction/2019%20834.194/2-2%20Hasu%20no%20Hana%20(Single%20Version).m4a
file:///home/user/audio/Sakanaction/2019%20834.194/2-3%20Eureka.m4a
file:///home/user/audio/Sakanaction/2019%20834.194/2-4%20Nylon%20no%20Ito.m4a
file:///home/user/audio/Sakanaction/2019%20834.194/2-5%20Chabashira.m4a
file:///home/user/audio/Sakanaction/2019%20834.194/2-6%20Wonderland.m4a
file:///home/user/audio/Sakanaction/2019%20834.194/2-7%20Sayonara%20wa%20Emotion.m4a
file:///home/user/audio/Sakanaction/2019%20834.194/2-8%20834.194.m4a
file:///home/user/audio/Sakanaction/2019%20834.194/2-9%20September%20(Sapporo%20Version).m4a
```

**Comment by Oleksiy-Yakovenko** [comment]
@niedzielski Maybe I'm missing something, but it feels like combining Playlist autosort with your option A would do just what you need. Unless of course you want to have a hybrid of manually and automatically sorted workflow.

**Comment by niedzielski** [comment]
@Oleksiy-Yakovenko, are you talking about this checkbox on each playlist? It doesn't seem to have any effect in this case at least. I tried appended with it on and off but the outputs were the same.

<img width="806" height="693" alt="Image" src="https://github.com/user-attachments/assets/c093ec01-2808-45e7-800a-d44fbe82ab10" />

There's also an option under the file browser plugin but maybe this is for that old deprecated browser. I don't see any effect with this either on or off.

<img width="1289" height="746" alt="Image" src="https://github.com/user-attachments/assets/71d431fd-71ff-43d8-bdfc-61066e155b51" />

**Comment by Oleksiy-Yakovenko** [comment]
>  are you talking about this checkbox on each playlist? 

Yes. And the way it works is that you need to sort playlist at least once during current session, before enabling the option.
After you do this -- it should remember that sorting option, and will autosort playlist the same way when files are added.

**Comment by niedzielski** [comment]
I just tried but can't seem to get it working on v1.10.2:

1. Disable autosort from prior session.
2. Add folder.
3. Sort by column (click twice for ascending).
4. Delete additions.
5. Enable autosort.
6. Add folder.

**Comment by Oleksiy-Yakovenko** [comment]
ok thanks for trying
I'll check it out -- this feature is a bit fragile, broke before.. maybe again

**Comment by Oleksiy-Yakovenko** [comment]
@niedzielski I verified, it actually works in the scenario you tried, but ONLY if you use custom title formatting for sorting, not just clicking a column :)

**Comment by niedzielski** [comment]
Thanks for the follow up! I just tried:

1. Add folder.
2. Edit -> Sort by -> Custom -> `%path%`.
3. Delete additions.
4. Add folder.

Step 2 works but the setting isn't retained for step 4.

To be honest, I'm unlikely to remember to go through these steps each session anyway but I appreciate the help!

**Comment by Oleksiy-Yakovenko** [comment]
You don't need to do it on every session. Once you enable it on a playlist - it's retained, and works even after restart.

**Comment by niedzielski** [comment]
That sounds promising! I'll keep fiddling with it but I can't seem to get it to work following these steps:

1. Add folder.
2. Edit -> Sort by -> Custom -> `%path%`.
3. Delete additions.
4. Add folder.

Step 2 works but the setting just isn't retained for step 4. I also tried `$filename(%path%)` and enabling playlist autosort before and after.
